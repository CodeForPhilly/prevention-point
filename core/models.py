from django.db import models
from django.utils import timezone
from enum import Enum
from core.participants.models import Participant
class Medication(models.Model):
    participant = models.ForeignKey(Participant, on_delete=models.CASCADE)
    medical_delivery = models.CharField(max_length=100)
    medication_name = models.CharField(max_length=100)
    ingestion_frequency = models.IntegerField()

class UrineDrugScreen(models.Model):
    participant = models.ForeignKey(Participant, on_delete=models.CASCADE)
    date_of_test = models.DateField()
    uds_temp = models.CharField(max_length=100, verbose_name="Urine Temperature")
    # TODO: does this belong here? if so, is it a Bool for if a PT was taken, or Bool for if pregnant?
    pregnancy_test = models.BooleanField(default=False)
    opiates = models.BooleanField(default=False)
    fentanyl = models.BooleanField(default=False)
    bup = models.BooleanField(default=False, verbose_name="Buprenorphine")
    coc = models.BooleanField(default=False, verbose_name="Cocaine")
    amp = models.BooleanField(default=False, verbose_name = "Amphetamine")
    m_amp = models.BooleanField(default=False, verbose_name = "Methamphetamine")
    thc = models.BooleanField(default=False, verbose_name = "THC")
    mtd = models.BooleanField(default=False, verbose_name = "Methadone")
    pcp = models.BooleanField(default=False, verbose_name = "PCP")
    bar = models.BooleanField(default=False, verbose_name = "Barbiturates")
    bzo = models.BooleanField(default=False, verbose_name = "Benzodiazepines")
    tca = models.BooleanField(default=False, verbose_name = "Tricyclic Antidepressants")
    oxy = models.BooleanField(default=False, verbose_name = "Oxycodone")

    def __str__(self):
        return 'UDS #%i for Participant #%i' % (self.id, self.participant_id)

class EmployeeRole(models.Model):
    role_value = models.CharField(max_length=30)

    def __str__(self):
        return '%s' % (self.role_value)

# TODO: seed with front desk, case manager, admin

class Employee(models.Model):
    first_name = models.CharField(max_length=50)
    last_name = models.CharField(max_length=100)
    role = models.ForeignKey(EmployeeRole, on_delete=models.CASCADE)

    def __str__(self):
        return '%s %s %s' % (self.first_name, self.last_name, self.role)

class CaseManagement(models.Model):
    participant = models.ForeignKey(Participant, on_delete=models.CASCADE)
    crs_seen = models.BooleanField(default=False)

class Appointment(models.Model):
    participant = models.ForeignKey(Participant, on_delete=models.CASCADE)

    cm_employee = models.ForeignKey(
            Employee, related_name='cm_%(class)s', on_delete=models.CASCADE
            )

    crs_employee = models.ForeignKey(
            Employee, related_name='crs_%(class)s', on_delete=models.CASCADE
            )

    appointment_timestamp = models.DateTimeField()
    crs_timestamp = models.DateTimeField()
    crs_seen = models.BooleanField(default=False)
    cm_timestamp = models.DateTimeField()
    cm_seen = models.BooleanField(default=False)
    hcv_status = models.BooleanField(default=False)

# TODO: we probably want to move the HCV data into its own table(s), similar to how UDS works

class BehavioralHealthNotes(models.Model):
    participant = models.ForeignKey(Participant, on_delete=models.CASCADE)
    employee = models.ForeignKey(Employee, on_delete=models.CASCADE)
    note_timestamp = models.DateTimeField()
    behavior_note = models.TextField()

class HCVNotes(models.Model):
    participant = models.ForeignKey(Appointment, on_delete=models.CASCADE)
    employee = models.ForeignKey(Employee, on_delete=models.CASCADE)
    note_timestamp = models.DateTimeField()
    hcv_note = models.TextField()

class Visit(models.Model):
    participant = models.ForeignKey(Participant, on_delete=models.CASCADE)
    created_at = models.DateTimeField(default=timezone.now)

class ServiceEvent(models.Model):
    class ServiceEventPurpose(Enum):
        ARRIVED = 'arrived'
        CALLED = 'called'
        SEEN = 'seen'
        LEFT = 'left'

    visit = models.ForeignKey(Visit, on_delete=models.CASCADE)
    # service = models.ForeignKey(Service, on_delete=models.CASCADE)
    purpose = models.CharField(
        max_length=10,
        choices=[(tag.name, tag.value) for tag in ServiceEventPurpose]
    )
