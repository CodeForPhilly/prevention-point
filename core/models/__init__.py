from .insurer import *
from .participant import *
from .program import *
from .service import *
from .program_service_map import *
from .visit import *
from .hcv_notes import *
from .medication import *
from .appointment import *
from .case_management import *
from .front_desk_event import *
from .urine_drug_screen import *
from .program_availability import *
from .behavioral_health_notes import *