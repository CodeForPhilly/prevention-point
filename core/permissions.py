from django.contrib.auth.models import User, Group
from rest_framework.permissions import DjangoModelPermissions, BasePermission

FRONT_DESK = 'front_desk'
CASE_MANAGER = 'case_manager'
ADMIN = 'admin'

class DjangoModelPermissions(DjangoModelPermissions):
    """
    The request is authenticated using `django.contrib.auth` permissions.
    See: https://docs.djangoproject.com/en/dev/topics/auth/#permissions
    It ensures that the user is authenticated, and has the appropriate
    `add`/`change`/`delete` permissions on the model.
    This permission can only be applied against view classes that
    provide a `.queryset` attribute.


    Prevention Point Note: We are extending this Class as to include permission checks on GET
    requests. You MUST use this class to protect GET requests to endpoints. The DRF class this 
    ineherits from does not protect GET requests
    """
    perms_map = {
        'OPTIONS': [],
        'HEAD': [],
        'GET': ['%(app_label)s.view_%(model_name)s'],
        'POST': ['%(app_label)s.add_%(model_name)s'],
        'PUT': ['%(app_label)s.change_%(model_name)s'],
        'PATCH': ['%(app_label)s.change_%(model_name)s'],
        'DELETE': ['%(app_label)s.delete_%(model_name)s'],
    
    }
