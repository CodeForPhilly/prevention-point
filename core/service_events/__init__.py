from .serializer import *
from .views import *
