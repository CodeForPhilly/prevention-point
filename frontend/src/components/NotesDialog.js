import React, { useContext, useEffect } from "react"
import { observer } from "mobx-react-lite"
import PropTypes from "prop-types"
import Dialog from "@material-ui/core/Dialog"
import DialogActions from "@material-ui/core/DialogActions"
import DialogContent from "@material-ui/core/DialogContent"
import DialogContentText from "@material-ui/core/DialogContentText"
import DialogTitle from "@material-ui/core/DialogTitle"
import TextField from "@material-ui/core/TextField"

import PrevPointButton from "./PrevPointButton"
import { QueueStoreContext } from "../stores/QueueStore"

const NotesDialog = observer(
  ({ visibleDialog, toggleVisibleDialog, queueData, id }) => {
    const queueStore = useContext(QueueStoreContext)

    const handleSubmit = (queueId, visitId) => {
      queueStore.patchVisit(queueId, visitId, {
        notes: queueStore.participantNotes,
      })
      toggleVisibleDialog()
    }

    useEffect(() => {
      queueStore.setParticipantNotes(queueStore.getNotes(queueData, id))
    }, [queueStore, queueStore.getNotes, queueData, id])

    return (
      <Dialog
        open={visibleDialog}
        onClose={toggleVisibleDialog}
        aria-labelledby="note-dialog-title"
        fullWidth
        maxWidth={"sm"}
      >
        <DialogTitle id="note-dialog-title">Participant Notes</DialogTitle>
        <DialogContent>
          <DialogContentText>Enter a note:</DialogContentText>
          <TextField
            id="notes"
            type="text"
            placeholder="Enter new note here"
            margin="dense"
            autoFocus
            fullWidth
            multiline
            onChange={e => queueStore.setParticipantNotes(e.target.value)}
            value={queueStore.participantNotes}
          />
          <DialogActions>
            <PrevPointButton color="default" onClick={toggleVisibleDialog}>
              Cancel
            </PrevPointButton>
            <PrevPointButton
              onClick={() =>
                handleSubmit(queueData, id, queueStore.participantNotes)
              }
            >
              Submit
            </PrevPointButton>
          </DialogActions>
        </DialogContent>
      </Dialog>
    )
  }
)

NotesDialog.propTypes = {
  visibleDialog: PropTypes.bool,
  toggleVisibleDialog: PropTypes.func,
  queueData: PropTypes.number,
  id: PropTypes.number,
}

export default NotesDialog
