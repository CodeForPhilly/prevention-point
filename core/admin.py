from django.contrib import admin
from core.models import Participant, Medication, CaseManagement, Appointment, Employee, BehavioralHealthNotes
from core.service_events.models import ServiceEvent
from core.urine_drug_screens.models import UrineDrugScreen

# Register your models here.
admin.site.register(Medication)
admin.site.register(UrineDrugScreen)
admin.site.register(CaseManagement)
admin.site.register(Appointment)
admin.site.register(Employee)
admin.site.register(BehavioralHealthNotes)
admin.site.register(ServiceEvent)

class BaseInline(admin.StackedInline):
    extra = 0

class UrineDrugScreenInline(BaseInline):
    model = UrineDrugScreen

class AppointmentInline(BaseInline):
    model = Appointment

class MedicationInline(BaseInline):
    model = Medication

class CaseManagement(BaseInline):
    model = CaseManagement

class ParticipantAdmin(admin.ModelAdmin):
    inlines = [
            AppointmentInline,
            MedicationInline,
            CaseManagement,
            UrineDrugScreenInline,
            ]

admin.site.register(Participant, ParticipantAdmin)
