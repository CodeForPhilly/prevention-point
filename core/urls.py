"""preventionpoint URL Configuration

The `urlpatterns` list routes URLs to views. For more information please see:
    https://docs.djangoproject.com/en/2.1/topics/http/urls/
Examples:
Function views
    1. Add an import:  from my_app import views
    2. Add a URL to urlpatterns:  path('', views.home, name='home')
Class-based views
    1. Add an import:  from other_app.views import Home
    2. Add a URL to urlpatterns:  path('', Home.as_view(), name='home')
Including another URLconf
    1. Import the include() function: from django.urls import include, path
    2. Add a URL to urlpatterns:  path('blog/', include('blog.urls'))
"""
from django.contrib import admin
from django.urls import include, path
from rest_framework import routers
from rest_framework_swagger.views import get_swagger_view
from rest_framework_simplejwt import views as jwt_views

from core.users import views as user_views
from core.urine_drug_screens import views as uds_views
from core.participants import views as participant_views
from core.visits import views as visits_views
from core.front_desk_events import views as front_events_views
from core.programs import views as programs_views
from core.services import views as services_views
from core.queue import views as queues_views
from core.program_availability import views as program_availability_views
from core.program_service_map import views as program_service_map_views
from core.insurers import views as insurer_views
from core.medications import views as medications_views

admin.site.site_header = "Prevention Point Philadelphia"

router = routers.DefaultRouter()
router.register(r"users", user_views.UserViewSet)
router.register(r"groups", user_views.GroupViewSet)
router.register(r"visits", visits_views.VisitViewSet)
router.register(r"uds", uds_views.UrineDrugScreenViewSet)
router.register(r"medications", medications_views.MedicationViewSet)
router.register(r"participants", participant_views.ParticipantViewSet)
router.register(r"front-desk-events", front_events_views.FrontDeskEventViewSet)
router.register(r"programs", programs_views.ProgramViewSet)
router.register(r"program-service-map", program_service_map_views.ProgramServiceMapView)
router.register(r"services", services_views.ServiceViewSet)
router.register(r"insurers", insurer_views.InsurerViewSet)


schema_view = get_swagger_view(title="PreventionPoint API")

urlpatterns = [
    path("api/", include(router.urls)),
    path(
        "api/token/", jwt_views.TokenObtainPairView.as_view(), name="token_obtain_pair"
    ),
    path(
        "api/token/refresh/", jwt_views.TokenRefreshView.as_view(), name="token_refresh"
    ),
    path("api/token/verify/", jwt_views.TokenVerifyView.as_view(), name="token_verify"),
    path(
        "api/programs/<int:program_id>/program-availability/",
        program_availability_views.ProgramAvailabilityViewSet.as_view({"get": "list"}),
    ),
    path(
        "api/programs/<int:program_id>/program-availability/<int:pk>/",
        program_availability_views.ProgramAvailabilityViewSet.as_view(
            {"put": "update"}
        ),
    ),
    path(
        "api/programs/<int:program_id>/queue/",
        queues_views.QueueViewSet.as_view({"get": "retrieve"}),
    ),
    # above only 'registers' the retrieve path. to use other CRUD actions, must pass that explicitly to .as_view in another register
    path("swagger/", schema_view),
    path("admin/", admin.site.urls),
]
