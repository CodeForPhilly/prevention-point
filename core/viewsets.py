from rest_framework import generics, viewsets
from rest_framework.permissions import IsAuthenticated
from core.permissions import DjangoModelPermissions

class ModelViewSet(viewsets.ModelViewSet):
    permission_classes = [DjangoModelPermissions, IsAuthenticated]
