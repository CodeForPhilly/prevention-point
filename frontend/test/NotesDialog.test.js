import React from "react"
//import { mount } from "enzyme"
import { createMount } from "@material-ui/core/test-utils"
import NotesDialog from "../src/components/NotesDialog"

const mockQueueData = 1
const mockId = 1
// temporarily turning off snapshot testing: the UI will continue to change based on the needs of Prevention Point and therefore snapshot testing is a little out of scope at this time. Would prefer reliable unit testing
xdescribe("Render the NotesDialog component", () => {
  const visibleDialog = true
  const toggleVisibleDialog = jest.fn()
  let wrapper

  const mount = createMount()

  beforeEach(() => {
    wrapper = mount(
      <NotesDialog
        visibleDialog={visibleDialog}
        toggleVisibleDialog={toggleVisibleDialog}
        queueData={mockQueueData}
        id={mockId}
      />
    )
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it("should match the snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should have a notes input", () => {
    expect(wrapper.exists("#notes")).toEqual(true)
  })

  it("should have a submit button", () => {
    expect(wrapper.exists("#submit")).toEqual(true)
  })

  it("should have a cancel button", () => {
    expect(wrapper.exists("#cancel")).toEqual(true)
  })
})
