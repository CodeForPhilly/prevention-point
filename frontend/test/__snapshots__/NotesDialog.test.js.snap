// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render the NotesDialog component should match the snapshot 1`] = `"<div role=\\"presentation\\" class=\\"MuiDialog-root\\" style=\\"position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;\\"><div class=\\"MuiBackdrop-root\\" aria-hidden=\\"true\\" style=\\"opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\"></div><div tabindex=\\"0\\" data-test=\\"sentinelStart\\"></div><div class=\\"MuiDialog-container MuiDialog-scrollPaper\\" style=\\"opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\" role=\\"none presentation\\" tabindex=\\"-1\\"><div class=\\"MuiPaper-root MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth MuiPaper-rounded\\" role=\\"dialog\\" aria-labelledby=\\"note-dialog-title\\"><div class=\\"MuiDialogTitle-root\\" id=\\"note-dialog-title\\"><h2 class=\\"MuiTypography-root MuiTypography-h6\\">Participant Notes</h2></div><div class=\\"MuiDialogContent-root\\"><p class=\\"MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary\\">Enter a note about this participant.</p><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth\\"><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth Mui-focused Mui-focused MuiInputBase-formControl MuiInput-formControl MuiInputBase-multiline MuiInput-multiline MuiInputBase-marginDense\\"><textarea rows=\\"1\\" aria-invalid=\\"false\\" id=\\"notes\\" placeholder=\\"Enter a note about this participant\\" class=\\"MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline MuiInputBase-inputMarginDense MuiInput-inputMarginDense\\" style=\\"height: 0px; overflow: hidden;\\"></textarea><textarea aria-hidden=\\"true\\" class=\\"MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline MuiInputBase-inputMarginDense MuiInput-inputMarginDense\\" readonly=\\"\\" tabindex=\\"-1\\" style=\\"visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;\\"></textarea></div></div><div class=\\"MuiDialogActions-root MuiDialogActions-spacing\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\" id=\\"cancel\\"><span class=\\"MuiButton-label\\">Cancel</span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary\\" tabindex=\\"0\\" type=\\"button\\" id=\\"submit\\"><span class=\\"MuiButton-label\\">Submit</span><span class=\\"MuiTouchRipple-root\\"></span></button></div></div></div></div><div tabindex=\\"0\\" data-test=\\"sentinelEnd\\"></div></div>"`;
